#**Exercise 3 Using LASSO Regression to Model Fat Sales Data**

In [0]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
alldata = pd.read_csv(open('drive/My Drive/Colab Notebooks/finalmaster-ratios.csv', 'rb'))

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
alldata.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

In [0]:
import pandas as pd
import pandas
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, accuracy_score

In [0]:
allvariablenames = list(alldata.columns.values)
listofallpredictors = allvariablenames[8:]


In [0]:
predictors = alldata[listofallpredictors] 

In [0]:
target = alldata['# Purchases']
len(listofallpredictors)   

182

In [0]:
#split data into train and test sets, with 30% retained for test

pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)

In [0]:
#LASSO Model
model = LassoLarsCV(precompute=False,cv=10)
pred = model.fit(pred_train,tar_train)

In [0]:
#build coefficent chart

predictors_model=pd.DataFrame(listofallpredictors)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.8559082469420836]
['B01001036' 2.505332140782658]
['B01001037' 0.8891816621940855]
['B01001038' 1.5316627258281275]
['B02001005' 0.4125156931568726]
['B13014026' 0.480047651813577]
['B13014027' 0.6979528571367356]
['B13016001' 875240656.6675868]
['B19001017' 1.4834309461653117]


#Question1
Explain above code doing.

####**line1**
"predictors_model=pd.DataFrame(listofallpredictors)" :Create a new table called predictors_model. We use pd.dataframe to convert the listofallpredictors from list to dataframe.
####**line2**
'predictors_model.columns = ['label']':We rename the 'predictirs_model' to 'label'.
####**line3**
"predictors_model['coeff'] = model.coef_": We add a new column called coeff and print out all the label and coefficient.
####line4-6
for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)
We use for loop to make all predictors coefficient bigger than zero.


#Question2

Link: https://www.socialexplorer.com/data/ACS2010/metadata/?ds=ACS10&table=B01001

*['B01001014' 0.8559082469420836]* Male in age 40-44 years old. One more male aged in 40-44 years old, we will sell 0.8559082469420836 unit more BoBo bars.


*['B01001036' 2.505332140782658]*  Female in age 30-34 years old. One more female aged in 30-34 years old, we will sell 2.505332140782658 unit more BoBo bars.


*['B01001037' 0.8891816621940855]*  Female in age 35-39 years old. One more female aged in 35-39 years old, we will sell 0.8891816621940855 unit more BoBo bars.

*['B01001038' 1.5316627258281275]*  Female in age 40-44 years old. One more female aged in 35-39 years old, we will sell 1.5316627258281275 unit more BoBo bars.


link: https://www.socialexplorer.com/data/ACS2018/metadata/?ds=ACS18&table=B02001
*['B02001005' 0.4125156931568726]* Asian alone. One Asian Alone, we will sell 0.4125156931568726 unit more BoBo bars.

link: https://www.socialexplorer.com/data/ACS2018/metadata/?ds=ACS18&table=B13014
*['B13014026' 0.480047651813577]* One more woman from 15 to 50 years who had a birth in the past 12 months with Bachelor degree, we will sell 0.480047651813577 unit more BoBo bars.

*['B13014027' 0.6979528571367356]* One more woman from 15 to 50 years who had a birth in the past 12 months with Graduate or Professional Degree, we will sell 0.6979528571367356 unit more BoBo bars.

*['B13016001' 875240656.6675868]* One more woman from 15 to 50 years who had a birth in the past 12 months, we will sell 875240656.6675868 unit more BoBo bars.

*['B19001017' 1.4834309461653117]* One more household income in the past 12 month with income $200,000 or more, we will sell 1.4834309461653117 unit more BoBo cars.

#Question3

If I had to report two census varibles, I will choose B01001036 Female in age 30-34 years old. And B13016001 Women from 15 to 50 years who had a birth in the past 12 months because these two have a high coefficient.

In [0]:
#2
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

train_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('testing data MSE')
print(train_error)
'''Testing MSE is bigger than train MSE. MSE means the average square difference between estimate value
and actual value and how close to the fitted line. The smaller MSE means the closer to the fits data. As
we can see in here training data MSE 22025.56 ia better than testing data MSE 41549.72.'''


training data MSE
22025.561527911726
testing data MSE
41549.71988734899


'Testing MSE is bigger than train MSE. MSE means the average square difference between estimate value\nand actual value and how close to the fitted line. The smaller MSE means the closer to the fits data. As\nwe can see in here training data MSE 24951.32 ia better than testing data MSE 32908.86.'

#Question4

In [0]:
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

rsquared_test=model.score(pred_test,tar_test)
print ('testing data R-square')
print(rsquared_test)

training data R-square
0.24001969075370644
testing data R-square
0.17585944255625585


As we can see here， testing data r-square 17.59% is smaller than training data r-square 24%. Therefore, the training data r_square is better than testing r square.

#Question5

Overall, the census data cannot accurately to predict the sales because as we can see the testing data MSE is too large to fit the data. And training data r-square is not bigger enough and it is hard to explain for the model so we don't have enough evidence to predict the sales.

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
22.198657392835337



#Question6

As we can see the baseline sells number is 22.1987, which means when other variable is zero like no female sell between 30-34 years old and so on, they will sell 22.1987 BoBo bar Sold.